client/src/client.rs

```diff
@@ -95,6 +95,7 @@ pub struct Client<B, E, Block, RA> where Block: BlockT {
 	storage_notifications: Mutex<StorageNotifications<Block>>,
 	import_notification_sinks: Mutex<Vec<mpsc::UnboundedSender<BlockImportNotification<Block>>>>,
 	finality_notification_sinks: Mutex<Vec<mpsc::UnboundedSender<FinalityNotification<Block>>>>,
+	all_blocks_notification_sinks: Mutex<Vec<mpsc::UnboundedSender<AllBlocksNotification<Block>>>>,
```

Comment:
I'm quite hesitant on adding another `Unbounded` channel here. This may become very unreliable, imagine a case where:
1. You have some non-trivial work to do on every block
2. The channel queues up a bunch of blocks, as you are not as fast processing them.
3. The state get's pruned in the meantime.

The whole scheme is then useless, cause even when we receive the notification the state is already lost. Also if the node is turned off the whole content of the channel is lost, so we will essentially miss running for some blocks.

If we want to make it super reliable and make sure that EACH and EVERY block is indexed/processed we need to come up with a more sophisticated mechanism.
I imagine something like LMAX-Disruptor scheme, where we:
1. Persist the state of indexing in some DB
2. Pin the state, up until the slower consumer processes it.
3. Proceed as fast as possible with fast consumers.

Reply:
> I'm quite hesitant on adding another Unbounded channel here.

Same here.